Change build scripts to be independent from the working directory

hardliner66 opened the pull request:
This change allows running the build scripts from any directory, making it more robust and less likely for people to use it wrong.

xslendix commented:
1. pushd/popd is a bash-specific which means you removed compatibility for a bunch of systems.
2. You should restore the previous PWD.
3. You could also make the Powershell scripts behave the same.

hardliner66 commented:
I removed pushd and switched to an approach using test and a subshell. This way it even returns to the old dir when there is an error.
I also wrapped the powershell scripts in a try-finally and use pushd/popd there as well.

xslendix commented on build/sync.sh:
```diff
@@ -1,12 +1,23 @@
+SCRIPT_DIR=$(realpath "$(dirname "$0")")
+SCRIPT_NAME=$(basename "$0")
+EXPECTED_DIR=$(realpath "$PWD")
+
+if test "${EXPECTED_DIR}" != "${SCRIPT_DIR}"
+then
+	( cd "$SCRIPT_DIR" || exit ; "./$SCRIPT_NAME" );
+	exit
+fi
```
Here, you can store the previous PWD:
```sh
PREV_PWD="$PWD"
if [ "${EXPECTED_DIR}" != "${SCRIPT_DIR}" ]; then
	cd "$SCRIPT_DIR" && "./$SCRIPT_NAME"
	cd "$PREV_DIR"
	exit
fi
```

xslendix commented on build/build-iso.sh:
```diff
@@ -2,7 +2,16 @@
+SCRIPT_DIR=$(realpath "$(dirname "$0")")
+SCRIPT_NAME=$(basename "$0")
+EXPECTED_DIR=$(realpath "$PWD")
+
+if test "${EXPECTED_DIR}" != "${SCRIPT_DIR}"
+then
+	( cd "$SCRIPT_DIR" || exit ; "./$SCRIPT_NAME" );
+	exit
+fi
```
Same thing as in sync.sh, you can store the previous PWD:
```sh
PREV_PWD="$PWD"
if [ "${EXPECTED_DIR}" != "${SCRIPT_DIR}" ]; then
	cd "$SCRIPT_DIR" && "./$SCRIPT_NAME"
	cd "$PREV_DIR"
	exit
fi
```

hardliner66 replied:
Same as https://github.com/Zeal-Operating-System/ZealOS/pull/57#discussion_r861807374.

xslendix commented on build/sync.ps1:
```diff
@@ -1,108 +1,117 @@
-﻿$choice = $args[0]
-
-function print_usage()
+﻿try
```
Are the `try` wrappers necessary? Why not do something similar to the shell scripts?

hardliner66 replied:
Yes. Otherwise the script wont return to the previous directory in case of an exception.

xslendix replied:
Yes, but you could spawn a new powershell instance to re-run the script in the correct directory. I personally think this is a much simpler and cleaner solution since it doesn't require as much of a drastic change.

hardliner66 replied:
It seems that you need the try/finally, otherwise a crash wont return you to the directory you where. I replicated the same behavior as with the shell script.

xslendix commented on build/build-iso.ps1:
```diff
@@ -1,66 +1,75 @@
-﻿if (!((Split-Path -Path (Get-Location) -Leaf) -eq "build"))
+﻿try
```
Same here, why is the `try` needed?

xslendix commented on build/build-iso.ps1:
```diff
@@ -1,66 +1,75 @@
+finally
+{
+	popd
+}
```
Could use a newline at the end of the file.

xslendix reviewed (commented):
I don't understand the need of `try` in the Powershell scripts, neither the indentation.
